[staging] ffmpeg: ffmpeg_4 → ffmpeg_5

fabianhjr opened the pull request:
###### Description of changes


###### Things done

Currently building and testing

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

fabianhjr committed:
ffmpeg: ffmpeg_4 → ffmpeg_5

fabianhjr commented:
```
Estimating rebuild amount by counting changed Hydra jobs (parallel=unset).
    161 x86_64-darwin
   2910 x86_64-linux
```

Atemu commented:
Oh, that's relatively tame as far as rebuilds go. We might not need a hydra jobset for that. I'll see whether I can run nixpkgs-reivew on that tomorrow.

Is there a good easy way to exclude already broken packages in nixpkgs-review?

fabianhjr commented:
Not that I know of :c

doronbehar commented:
Speaking of cleanup, [ffmpeg 6 is out](https://git.ffmpeg.org/gitweb/ffmpeg.git/blob/refs/heads/release/6.0:/RELEASE_NOTES)

I tried to package it with:

```diff
diff --git c/pkgs/development/libraries/ffmpeg/6.nix i/pkgs/development/libraries/ffmpeg/6.nix
new file mode 100644
index 00000000000..37bdf5b060f
--- /dev/null
+++ i/pkgs/development/libraries/ffmpeg/6.nix
@@ -0,0 +1,4 @@
+import ./generic.nix {
+  version = "6.0";
+  sha256 = "sha256-RVbgsafIbeUUNXmUbDQ03ZN42oaUo0njqROo7KOQgv0=";
+}
diff --git c/pkgs/top-level/all-packages.nix i/pkgs/top-level/all-packages.nix
index 1ddec927d95..fedc4500268 100644
--- c/pkgs/top-level/all-packages.nix
+++ i/pkgs/top-level/all-packages.nix
@@ -19762,6 +19762,18 @@ with pkgs;
     ffmpegVariant = "full";
   };
 
+  ffmpeg_6 = callPackage ../development/libraries/ffmpeg/6.nix {
+    inherit (darwin.apple_sdk.frameworks)
+      Cocoa CoreServices CoreAudio CoreMedia AVFoundation MediaToolbox
+      VideoDecodeAcceleration VideoToolbox;
+  };
+  ffmpeg_6-headless = ffmpeg_6.override {
+    ffmpegVariant = "headless";
+  };
+  ffmpeg_6-full = ffmpeg_6.override {
+    ffmpegVariant = "full";
+  };
+
   # Aliases
   # Please make sure this is updated to the latest version on the next major
   # update to ffmpeg
```

But it fails to detect the `cuvid` dependency:

```
ERROR: cuvid requested, but not all dependencies are satisfied: ffnvcodec
```

Although according to [this upstream commit](https://git.ffmpeg.org/gitweb/ffmpeg.git/commitdiff/85051febc62e396cbe46c55d674fa809c890d3f9?hp=c775410f318353941dc980b78f2ed4807ee66046) and the fact we add `nv-codec-headers-9.1.23.1` as a dependency.

Atemu commented:
Thanks for taking a look @doronbehar but could you open a separate issue on that (and ping me)?

Atemu commented:
```

243 packages marked as broken and skipped:
_7kaa alfaview andyetitmoves anytype apmplanner2 aqemu archivebox assaultcube aws-workspaces bespokesynth-with-vst2 bitwig-studio bitwig-studio1 bitwig-studio2 bitwig-studio3 bitwig-studio4 bootstrap-studio calligra citrix_workspace citrix_workspace_23_02_0 cudaPackages.nsight_compute davinci-resolve diffoscope docear dropbox dropbox-cli dwarf-fortress dwarf-fortress-packages.dwarf-fortress dwarf-fortress-packages.dwarf-fortress-full dwarf-fortress-packages.dwarf-fortress-original dwarf-fortress-packages.dwarf-fortress_0_43_05 dwarf-fortress-packages.dwarf-fortress_0_44_05 dwarf-fortress-packages.dwarf-fortress_0_44_09 dwarf-fortress-packages.dwarf-fortress_0_44_10 dwarf-fortress-packages.dwarf-fortress_0_44_11 dwarf-fortress-packages.dwarf-fortress_0_44_12 dwarf-fortress-packages.dwarf-fortress_0_47_01 dwarf-fortress-packages.dwarf-fortress_0_47_02 dwarf-fortress-packages.dwarf-fortress_0_47_04 dwarf-fortress-packages.dwarf-fortress_0_47_05 dwarf-fortress-packages.dwarf-therapist dwarf-therapist eagle elementary-planner everspace exactaudiocopy feishu flexoptix-app forge frogatto frostwire gimpPlugins.gmic gmic-qt gnome.gnome-todo goldendict google-chrome google-chrome-beta google-chrome-dev gpu-screen-recorder gpu-screen-recorder-gtk gscrabble home-assistant-component-tests.nest hottext hqplayer-desktop hqplayerd ib-controller ib-tws insync-v3 intensity-normalization itch javacard-devkit jre8Plugin jrePlugin keen4 keet kexi koboredux kodiPackages.steam-launcher libsForQt5.full libsForQt5.kreport libsForQt5.qt5.full linuxKernel.packages.hardkernel_4_14.virtualbox linuxKernel.packages.linux_5_4_hardened.virtualbox linuxPackages_5_4_hardened.virtualbox linuxPackages_hardkernel_latest.virtualbox lispPackages_new.sbclPackages.clinch-classimp lispPackages_new.sbclPackages.nyxt_slash_application_slash_tests lispPackages_new.sbclPackages.nyxt_slash_benchmark lispPackages_new.sbclPackages.nyxt_slash_documentation lispPackages_new.sbclPackages.nyxt_slash_gi-gtk lispPackages_new.sbclPackages.nyxt_slash_gi-gtk-application lispPackages_new.sbclPackages.nyxt_slash_gi-gtk_slash_tests lispPackages_new.sbclPackages.nyxt_slash_gtk lispPackages_new.sbclPackages.nyxt_slash_gtk-application lispPackages_new.sbclPackages.nyxt_slash_tests lispPackages_new.sbclPackages.nyxt_slash_version lispPackages_new.sbclPackages.sdl2_slash_examples lunar-client lutris mailnag mailnagWithPlugins makemkv mate.caja-dropbox mathematica mathematica-cuda mathpix-snipping-tool maui-shell megasync mendeley minigalaxy minizincide molotov nasc neo4j-desktop netflix notable notion-app-enhanced nxengine-evo obs-studio-plugins.obs-hyperion openjfx19 openmw-tes3mp opentoonz openxray opera oraclejdk oraclejdk8 oraclejre oraclejre8 osu-lazer osu-lazer-bin p4v palemoon pantheon.elementary-tasks pantheon.evince pantheon.file-roller parsec-bin pinegrow pinegrow6 plasma5Packages.full plasma5Packages.kreport plasma5Packages.qt5.full playonlinux plexamp polypane processing protontricks pureref python310Packages.bambi python310Packages.caffe python310Packages.caffeWithCuda python310Packages.imgaug python310Packages.intensity-normalization python310Packages.mask-rcnn python310Packages.pywick python310Packages.rl-coach python310Packages.rlax qt5.full qt5Full quake3demo rambox reaper remnote retrofe roon-bridge roon-server runescape rymcast saleae-logic-2 signumone-ks sky slack sm64ex sm64ex-coop snes9x-gtk space-cadet-pinball spotify spotifywm sqldeveloper starsector steam steam-run steam-small steam-tui steamcmd sunvox tdlib-purple teamspeak_client teamviewer tetrd timeular tonelib-zoom transcribe trezor-suite trillian-im trojita tumpa ue4demos.black_jack ue4demos.blueprint_examples_demo ue4demos.card_game ue4demos.effects_cave_demo ue4demos.elemental_demo ue4demos.landscape_mountains ue4demos.matinee_demo ue4demos.mobile_temple_demo ue4demos.realistic_rendering ue4demos.reflections_subway ue4demos.scifi_hallway_demo ue4demos.shooter_game ue4demos.strategy_game ue4demos.stylized_demo ue4demos.swing_ninja ue4demos.tappy_chicken ue4demos.vehicle_game uhk-agent uhk-udev-rules unig

3 packages blacklisted:
appimage-run-tests nixos-install-tools tests.nixos-functions.nixos-test

142 packages failed to build:
acoustidFingerprinter ardour ardour_6 attract-mode bino3d bitcoin-unlimited bombono ccextractor citra-canary citra-nightly devede digikam dogecoin doodle dvdstyler electricsheep fast-cli febio-studio fluxus gmtk gnome-decoder gnomeExtensions.easyScreenCast gnome_mplayer gnunet gnunet-gtk gopro guvcview harvid haskellPackages.ffmpeg-light haskellPackages.mplayer-spot hedgewars hollywood home-assistant-component-tests.reolink hydron kmplayer kotatogram-desktop kotatogram-desktop-with-webkit libextractor liblinphone libretro.ppsspp libsForQt5.libqtav linphone lispPackages_new.sbclPackages.cepl_dot_sdl2-image lispPackages_new.sbclPackages.cepl_dot_sdl2-ttf lispPackages_new.sbclPackages.easing-demo lispPackages_new.sbclPackages.nyxt-ubuntu-package lispPackages_new.sbclPackages.sdl2-image lispPackages_new.sbclPackages.sdl2-mixer lispPackages_new.sbclPackages.sdl2-ttf lispPackages_new.sbclPackages.sdl2-ttf-examples lispPackages_new.sbclPackages.sketch lispPackages_new.sbclPackages.sketch-examples loudgain manaplus mediastreamer mediastreamer-openh264 miniplayer moc monado mplayer msilbc musly napari octavePackages.video olive-editor openai-whisper openclonk openhmd openrw oven-media-engine pangolin pcsxr performous pivx ppsspp ppsspp-qt ppsspp-sdl-wayland puppeteer-cli python310Packages.dask-image python310Packages.dm-sonnet python310Packages.ffmpeg-python python310Packages.pims python310Packages.pytest-playwright python310Packages.reolink python310Packages.reolink-aio python310Packages.shap python310Packages.spectral-cube python310Packages.stytra python310Packages.sunpy python310Packages.tensorflow-datasets python311Packages.ffmpeg-python python311Packages.homeassistant-stubs python311Packages.humblewx python311Packages.nocturne python311Packages.pyautogui python311Packages.pygame python311Packages.pygame-gui python311Packages.pygetwindow python311Packages.pyrect python311Packages.pytest-playwright python311Packages.pythonefl python311Packages.pytmx python311Packages.reolink python311Packages.reolink-aio python311Packages.wxPython_4_0 python311Packages.wxPython_4_1 qstopmotion quodlibet quodlibet-full quodlibet-without-gst-plugins quodlibet-xine quodlibet-xine-full r128gain retroarchFull reveal-md rtabmap shadered sharedown simplescreenrecorder sonic-pi spek spring springLobby stepmania sumo syncterm taler-exchange taler-merchant tenacity timeline tvheadend ultrastardx untrunc-anthwlock vhs vulkan-cts webcamoid wtwitch xine-lib xine-ui xjadeo ytmdl zombietrackergps

2497 packages built:
ChowKick CuboCore.corepdf CuboCore.coretoppings EmptyEpsilon MIDIVisualizer PageEdit SDL2 SDL2_gfx SDL2_image SDL2_image_2_0_5 SDL2_mixer SDL2_net SDL2_sound SDL2_ttf SDL2_ttf_2_0_15 SDL_audiolib SDL_compat SDL_gpu Sylk _1oom _20kly _90secondportraits aaxtomp3 abbaye-des-morts adapta-gtk-theme adl aegisub ahoviewer aitrack alass alephone alfis allegro allegro5 almanah alpine-make-vm-image alsa-scarlett-gui altair alure alure2 amarok amberol anbox ani-cli anime-downloader anki anki-bin ankisyncd antimicrox antsimulator apache-directory-studio apostrophe appimage-run aravis arcanPackages.all-wrapped arcanPackages.arcan arcanPackages.arcan-wrapped arcanPackages.cat9-wrapped arcanPackages.durden-wrapped ffmpeg-full arcanPackages.pipeworld-wrapped arcanPackages.prio-wrapped arcanPackages.xarcan ares armcord arx-libertatis asc-key-to-qr-code-gif asciidoc-full-with-plugins ashpd-demo astroid astromenace atanks audacious audacious-plugins audacity audiality2 audio-recorder augustus authenticator auto-multiple-choice bada-bib balsa banking baobab baresip basiliskii bazarr beamerpresenter beamerpresenter-poppler beekeeper-studio beets beets-unstable beneath-a-steel-sky bespokesynth betaflight-configurator betterdiscord-installer birdfont bitcoin bitcoin-abc blackbox-terminal blackshades blanket blender blender-hip blobby blobwars blockattack bloomrpc blueprint-compiler bochs bookworm botamusique bottles bottles-unwrapped brasero brasero-original brave breath-theme broken-sword-25 brutalmaze brutespray bsnes-hd bubblemail bucklespring bucklespring-libinput budgie.budgie-screensaver bugdom buttercup-desktop byzanz bzflag cadence cage cagebreak calamares calamares-nixos calibre calls cambalache cantata caprine-bin capture cardboard cargo-tauri cataclysm-dda cataclysm-dda-git catt cavalier cawbird cdogs-sdl celluloid cemu cen64 cfdg cgui chatty chiaki chirp chocolateDoom chromaprint chrysalis ciano cider cinnamon.cinnamon-common cinnamon.cinnamon-control-center cinnamon.cinnamon-gsettings-overrides cinnamon.cinnamon-screensaver cinnamon.muffin cinnamon.nemo-fileroller cinnamon.nemo-with-extensions cinnamon.pix cinnamon.xreader cinny-desktop citations clapper classicube claws-mail clementine clickclack clipgrab cloud-init cloud-utils cmus cog colima colobot colord-gtk4 commandergenius contrast corrscope cot cozy cozy-drive cplay-ng craftos-pc crawlTiles crispyDoom crow-translate crowbar crypto-org-wallet csfml csound-qt ctx curseradio cutemarked-ng cutter cyanrip czkawka d-spy dablin darkplaces darktable dbeaver ddnet deepin.deepin-album deepin.deepin-camera deepin.deepin-image-viewer deepin.deepin-reader deepin.deepin-voice-note deepin.image-editor deja-dup denaro denemo desmume devdocs-desktop devilutionx dhewm3 dialect digitalbitbox dino dippi dmlive dolphin-emu dolphin-emu-primehack dosbox dosbox-staging dr14_tmeter drascula-the-vampire-strikes-back dreamweb droidcam dsp duckstation dumb dvd-slideshow dwl dxvk_2 dxx-rebirth dynamic-wallpaper eaglemode eartag easyeffects easyocr easyrpg-player eclipse-mat eclipses.eclipse-committers eclipses.eclipse-cpp eclipses.eclipse-java eclipses.eclipse-jee eclipses.eclipse-modeling eclipses.eclipse-platform eclipses.eclipse-rcp eclipses.eclipse-scala-sdk eclipses.eclipse-sdk ecwolf eduke32 eiciel electron-mail electronplayer elements emacsPackages.ac-mozc emacsPackages.mozc emacsPackages.mozc-cand-posframe emacsPackages.mozc-im emacsPackages.mozc-popup emacsPackages.mozc-temp emacsPackages.telega emblem emuflight-configurator emulationstation emulsion-palette endeavour endgame-singularity endless-sky enigma enlightenment.econnman enlightenment.ecrire enlightenment.efl enlightenment.enlightenment enlightenment.ephoto enlightenment.evisum enlightenment.rage enlightenment.terminology eolie eos-installer ephemeral epiphany eternity evince evolution evolution-data-server evolution-data-server-gtk4 evolution-ews evolutionWithPlugins extremetuxracer exult eyedropper ezquake facedetect factor-lang faircamp famistudio farstream faudio faust2jack faustlive fceux fcitx-engines.libpinyin fcitx5-chinese-a

```

Skipped wine (works) and chromium (unknown). Ignore the chromium-unwrapped failure, I stopped that intentionally because I didn't want to build chromium. I'll just set it to ffmpeg_4 too and let the chromium maintainers upgrade it themselves if they feel like it.

I'll also mark entirely broken packages as such.

Atemu commented:
Wait hold on, that list is broken. I'll need to get tmux to copy and paste properly somehow after eating. The list of broken packages is correct though.

Atemu committed:
treewide: make ffmpeg_4-dependant packages depend on ffmpeg_4

These packages were identified as broken when ffmpeg = ffmpeg_5.

Tested _only_ by building them. If you are a maintainer of one of these
packages, please test whether your package still works.

Atemu committed:
tree-wide: mark broken packages as such

Found while changing ffmpeg from 4 to 5; these all depend on ffmpeg either
directly or transitively.

Atemu commented:
Please pull https://github.com/Atemu/nixpkgs/tree/ffmpeg_5

fabianhjr commented:
:thinking: there is a merge conflict on the merge-base of `staging` and `master`.

I could rebase to either, will rebase to staging for the time being since that is the merge target.

fabianhjr commented:
```
 pkgs/development/python-modules/pygame/default.nix
pkgs/development/python-modules/shap/default.nix
pkgs/games/manaplus/default.nix 
```

fabianhjr commented:
Conflict resolved, rebased to staging (no longer on merge-base of staging and master)

doronbehar commented:
@ofborg eval

Atemu commented:
I forgot I still had some TODOs and unfortunately I lost my scratch buffer to a restart. It was only a few packages and I'm running nixpkgs-review on my tree to find them again and do some final QA. One was `haskellPackages.ffmpeg-light` which we'll need to somehow get on ffmpeg_4 and another was sonic-pi which doesn't compile on my machine but doesn't directly depend on ffmepg in any way?

I can no longer effectively review changes done now on staging; too many rebuilds. I'm not sure we really need a hydra jobset for this anymore though.

fabianhjr commented:
seems like the staging-main conflict has been resolved.

Rebased to the merge-base of both again.

SuperSandro2000 commented on nixos/modules/config/no-x-libs.nix:
```diff
@@ -30,8 +30,8 @@ with lib;
       dbus = super.dbus.override { x11Support = false; };
-      ffmpeg_4 = super.ffmpeg_4.override { ffmpegVariant = "headless"; };
-      ffmpeg_5 = super.ffmpeg_5.override { ffmpegVariant = "headless"; };
+      ffmpeg_4 = super.ffmpeg_4-headless;
```
please revert https://github.com/NixOS/nixpkgs/commit/9e3e928ad45fa9df11280be4ff581c4c11992e21

fabianhjr replied:
Removed the commit

SuperSandro2000 reviewed (changes requested):
I assume all the meta.broken changes are related to ffmpeg? Can you add a short comment to all of them with that information. Otherwise no one remembers why we marked those packages broken.

fabianhjr commented:
> I assume all the meta.broken changes are related to ffmpeg?

Likely broken before this change:

```
    Found while changing ffmpeg from 4 to 5; these all depend on ffmpeg either
    directly or transitively.

 pkgs/applications/blockchains/bitcoin-unlimited/default.nix | 2 +-
 pkgs/applications/blockchains/dogecoin/default.nix          | 1 +
 pkgs/applications/blockchains/pivx/default.nix              | 2 +-
 pkgs/applications/gis/zombietrackergps/default.nix          | 1 +
 pkgs/applications/graphics/fluxus/default.nix               | 1 +
 pkgs/applications/graphics/gnome-decoder/default.nix        | 1 +
 pkgs/applications/terminal-emulators/syncterm/default.nix   | 3 ++-
 pkgs/applications/video/rtabmap/default.nix                 | 1 +
 pkgs/desktops/gnome/extensions/EasyScreenCast/default.nix   | 2 +-
 pkgs/development/libraries/openhmd/default.nix              | 1 +
 pkgs/development/python-modules/apache-beam/default.nix     | 1 +
 pkgs/development/python-modules/pims/default.nix            | 1 +
 pkgs/development/python-modules/pydevd/default.nix          | 1 +
 pkgs/development/python-modules/python-efl/default.nix      | 1 +
 pkgs/development/python-modules/spectral-cube/default.nix   | 2 +-
 pkgs/development/python-modules/sunpy/default.nix           | 1 +
 pkgs/development/python-modules/wxPython/4.0.nix            | 1 +
 pkgs/development/python-modules/wxPython/4.1.nix            | 1 +
 pkgs/development/tools/shadered/default.nix                 | 1 +
 pkgs/games/openclonk/default.nix                            | 1 +
 pkgs/games/spring/default.nix                               | 1 +
```

 - bitcoin-unlimited: broken since 2023-01-14 / last build 2022-12-30
   https://hydra.nixos.org/build/210662522
 - dogecoin: broken since 2023-01-14 / last build 2023-01-07
   https://hydra.nixos.org/build/210775530
 - pivx: broken since 2020-12-11 / last build 2020-12-06
   https://hydra.nixos.org/build/210831321
 - zombietrackergps: broken since 2023-02-13 / last build 2023-02-12
   https://hydra.nixos.org/build/211324850
 - fluxus: broken since 2023-02-13 / last build 2023-02-06
   https://hydra.nixos.org/build/211234819
 - etc

Atemu commented:
@SuperSandro2000 the packages were already broken; that's unrelated to ffmpeg. It's just that I noticed they were broken because they depend on ffmpeg transitively which causes them to show up as "changed" in nixpkgs-review of course.

fabianhjr commented:
Rebasing, give me a minute

fabianhjr commented:
This was the conflict:

```
--- a/pkgs/applications/gis/zombietrackergps/default.nix
+++ b/pkgs/applications/gis/zombietrackergps/default.nix
@@@ -62,6 -62,6 +62,11 @@@ mkDerivation rec 
      license = licenses.gpl3Plus;
      maintainers = with maintainers; [ sohalt ];
      platforms = platforms.linux;
++<<<<<<< HEAD
 +    broken = true;  # doesn't build with latest Marble
++||||||| parent of 4eafc2c0c16 (tree-wide: mark broken packages as such)
++=======
+     broken = true;
++>>>>>>> 4eafc2c0c16 (tree-wide: mark broken packages as such)
    };
  }
```

Atemu commented:
We can also pull out the brokeness markings into a separate PR. I've got some more at https://github.com/atemu/nixpkgs/tree/ffmpeg_5.

vcunat commented:
I think this is ready from mass-rebuild perspective.  Further tweaks can come later.

ESTHERTERSEN commented:
Start2

ESTHERTERSEN commented:
> Start2

Arg-2

K900 commented:
Sorry?
